packages/guides/src/Compiler/NodeTransformers/SectionCreationTransformer.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @link https://phpdoc.org
+ */
+
+namespace phpDocumentor\Guides\Compiler\NodeTransformers;
+
+use phpDocumentor\Guides\Compiler\CompilerContext;
+use phpDocumentor\Guides\Compiler\NodeTransformer;
+use phpDocumentor\Guides\Nodes\DocumentNode;
+use phpDocumentor\Guides\Nodes\Node;
+use phpDocumentor\Guides\Nodes\SectionNode;
+use phpDocumentor\Guides\Nodes\TitleNode;
+
+use function array_pop;
+use function count;
+use function end;
+
+use const PHP_INT_MAX;
+
+/** @implements NodeTransformer<Node> */
+final class SectionCreationTransformer implements NodeTransformer
+{
+    /** @var SectionNode[] $sectionStack */
+    private array $sectionStack = [];
+
+    public function enterNode(Node $node, CompilerContext $compilerContext): Node
+    {
+        if (!$compilerContext->getShadowTree()->getParent()?->getNode() instanceof DocumentNode) {
+            return $node;
+        }
+
+        if (!$node instanceof TitleNode) {
+            $lastSection = end($this->sectionStack);
+            if ($lastSection instanceof SectionNode) {
+                $lastSection->addChildNode($node);
+            }
+        }
+
+        return $node;
+    }
+
+    public function leaveNode(Node $node, CompilerContext $compilerContext): Node|null
+    {
+        if (!$compilerContext->getShadowTree()->getParent()?->getNode() instanceof DocumentNode) {
+            return $node;
+        }
+
+        if ($node instanceof SectionNode) {
+            return $node;
+        }
+
+        if (count($this->sectionStack) === 0 && !$node instanceof TitleNode) {
+            return $node;
+        }
+
+        if (count($this->sectionStack) > 0 && $compilerContext->getShadowTree()->isLastChildOfParent()) {
+            $lastSection = end($this->sectionStack);
+            while ($lastSection?->getTitle()->getLevel() > 1) {
+                $lastSection = array_pop($this->sectionStack);
+            }
+
+            return $lastSection;
+        }
+
+        if ($node instanceof TitleNode) {
```

Comment:
Inverse this logic would make sense to me, as this would make it more readable what happens, only when the current node is a TitleNode we do more complex stuff, This can even be moved 1 if statement up? 
```suggestion
        if ($node instanceof TitleNode === false) {
            return null;
        }
```